lsm_tree.c

```diff
@@ -0,0 +1,333 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <stdbool.h>
+#include "lsm_tree.h"
+#include "memtable.h"
+#include "custom_io.h"
+
+// prototypes for static functions here
+static char** init_segments();
+static char* compact(char **segment_files);
+static char* merge_files(char *segment_file1, char *segment_file2);
+static void merge_lines(char *line_a, char *line_b, FILE *new_fp, bool *incr_a,
+bool *incr_b);
+static char* search_segments(char **segment_files, int full_segments, int key);
+static char* search_segment_file(FILE *segment_ptr, int key);
+static void free_segments(char **segments);
+
+/* Creates an LSM Tree for the program to use, initializing
+ * everything properly */
+LSM_Tree* init_lsm_tree() {
+	LSM_Tree *lsm_tree = (LSM_Tree*) malloc(sizeof(LSM_Tree));
+	if (lsm_tree == NULL) {
+		printf("Allocation of memory for LSM Tree failed.\n");
+		return NULL;
+	}
+
+	Memtable *memtable = init_memtable();
+	if (memtable == NULL) {
+		free(lsm_tree);
+		return NULL;
+	}
+
+	char **segments = init_segments();
+	if (segments == NULL) {
+		free(lsm_tree);
+		free(memtable);
+		return NULL;
+	}
+
+	lsm_tree->memtable = memtable;
+	lsm_tree->segments = segments;
+	lsm_tree->full_segments = 0;
+	return lsm_tree;
+}
+
+/* Handles the submission provided by user; returns 0 if all succeeds,
+ * otherwise returns -1.  */
+int handle_submission(LSM_Tree *lsm_tree, Submission *submission) {
+	if (submission->action == ADD) {
```

Comment:
For each action might want to add some small validation. Off the top of my head I'm thinking for ```ADD``` validate the value is not the tombstone ```string``` and for ```DELETE``` validate value is null

Reply:
Not sure what you mean the delete step, but changed the `ADD` to check for tombstone.